sceptre/resolvers/__init__.py

```diff
@@ -3,30 +3,53 @@
 import logging
```

Comment:
This module is where the bulk of the real change for this pull request comes from.

## Here's a summary of what lies below
1. ResolvableProperty has been turned into a base class with two subclasses:
    * ResolvableContainerProperty - This is mostly the same as what ResolvableProperty was previously; It is intended to resolve dicts and lists of resolvers. 
    * ResolvableValueProperty - This is brand new. It is a resolver that resolves to a single value, like an int, string, or other non-container value. This was required to support resolvers for non-containers like `template_bucket_name`, `iam_role`, and `role_arn`.
2. The ResolvableContainerProperty has been enhanced so that we no longer need the weird second-layer property on the Stack class for sceptre_user_data that resolves the ResolveLater objects. The ResolvableContainerProperty now handles those "deferred resolutions" properly.
3. There was a bug in the _no_recursive_get method that would block the same property from being accessed on a _different_ stack. _This was one (of several) issues blocking iam_role from being resolvable_. In a nutshell, this was because the stack_output resolver uses the iam_role of the _targeted_ stack. If we resolve stack A's iam_role from stack B's 's output, technically the iam_role property (which is an attribute on the stack _class_, not instance) is accessed twice. I needed to allow the same property to be accessed on two different stacks at the same time (since there should be no conflict with that).
4. Resolvers that return `None` in the ResolvableContainerProperty will have their key/value pair _removed_ from the container. This is what I've described as "resolving to nothing". This is described in documentation section of this PR. _This part was essential for eliminating inherited dependencies from project dependencies._
5. Resolvers (and all containers) are _cloned_ when they are assigned to the stack. This is really important because there was a bug where a resolver that was inherited from the stack group config would otherwise be the _same_ object in memory and would end up having one of several stacks assigned to it, so all the other stacks would accidentally be modifying other stacks.
6. It is possible to enter a global state (using a context manager) that allows Resolvers that raise errors to resolve to a placeholder value. This is explicitly opt-in and will only be explicitly enabled on commands that don't actually deploy stacks, such as "generate", "validate", and "diff". _This is required so that these commands don't fail when there are resolvers that reference stacks that aren't deployed yet._